### PR TITLE
Dw 3603 create analytical dataset generation emr cluster

### PR DESCRIPTION
**Pull request by UdayChokkam**

### CODE REVIEW

File: resources.tf [hunk 1]
```diff
@@ -247,6 +461,27 @@ resource "aws_security_group_rule" "analytical_dataset_generation_ingress" {
...
+  from_port                = 443
+  protocol                 = "tcp"
+  security_group_id        = aws_security_group.analytical_dataset_generation.id
+  to_port                  = 443
+  type                     = "egress"
+  source_security_group_id = data.terraform_remote_state.internal_compute.outputs.vpc.interface_vpce_sg_id
+
+}
+
+resource "aws_security_group_rule" "ingress_https_vpc_endpoints_from_emr" {
```
Comment (ChrisScottThomas) [inline resources.tf]: I think this needs to be done in the vpc enpoint to accept the connection from our SG. @mattburgess ?

File: resources.tf [hunk 2]
```diff
@@ -207,6 +327,100 @@ data "aws_iam_policy_document" "analytical_dataset_write_s3" {
   }
 }
 
+resource "aws_iam_role_policy_attachment" "amazon_ssm_managed_instance_core" {
```
Comment (ChrisScottThomas) [inline resources.tf]: Can this not be added to the current iam role?

File: emr.tf [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  name                              = local.emr_cluster_name
+  release_label                     = var.emr_release_label
+  applications                      = var.emr_applications
+  termination_protection            = var.termination_protection
+  keep_job_flow_alive_when_no_steps = var.keep_flow_alive
+  //TODO The below is need for transparent encryption/decryption when insecure DKS is set up
+  //security_configuration            = aws_emr_security_configuration.emrfs_em.id
+  service_role                      = aws_iam_role.analytical_dataset_generator.arn
+  //TODO which is the right bucket for logs , publish bucket??
+  log_uri                           = format("s3n://%s/logs/", aws_s3_bucket.published.id)
```
Comment (mattburgess) [inline emr.tf]: This should point at the logstore bucket managed by aws-security-tools. See https://github.ucds.io/dip/aws-security-tools/blob/master/logstore.tf#L85-L89

File: emr-setup.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
+
+set -e
```
Comment (ChrisScottThomas) [inline emr-setup.sh:3]: you want these enabled after testing?

File: emr.tf [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  configurations_json = templatefile(format("%s/configuration.json", path.module), {
+    logs_bucket_path = format("s3://%s/logs", data.terraform_remote_state.security-tools.outputs.logstore_bucket.id)
+    data_bucket_path = format("s3://%s/data", aws_s3_bucket.published.id)
+    //TODO this path needs to be taken from the output of aws-ingestion
+    hbase_root_path    = format("s3://%s/business-data/single-topic-per-table-hbase", data.terraform_remote_state.ingest.outputs.s3_buckets.input_bucket)
+    hive_external_path = format("s3://%s/hive/external", aws_s3_bucket.published.id)
+    proxy_host         = data.terraform_remote_state.internet_egress.outputs.internet_proxy.dns_name
+  })
+
+  step {
```
Comment (ChrisScottThomas) [inline emr.tf:59]: Can we not repeat line 50 for a steps.json ?

File: variables.tf [hunk 6]
```diff
@@ -0,0 +1,31 @@
+variable "emr_release_label" {
+  description = "Version of AWS EMR to deploy with associated applicatoins"
+  default     = "emr-5.29.0"
```
Comment (ChrisScottThomas) [inline variables.tf:3]: I think this should match the ingest-cluster.  Which is emr-5.24.1.

File: resources.tf [hunk 7]
```diff
@@ -166,10 +166,130 @@ data "aws_iam_policy_document" "analytical_dataset_write_s3" {
       "s3:GetObject*",
       "s3:DeleteObject*",
       "s3:PutObject*",
+
+      "acm:ExportCertificate",
+      "cloudwatch:*",
+      "dynamodb:*",
+      "ec2:Describe*",
+      "elasticmapreduce:Describe*",
```
[the remark this answers is not in the document]
Comment (UdayChokkam) [reply resources.tf]: Done

File: resources.tf [hunk 8]
```diff
@@ -166,10 +166,130 @@ data "aws_iam_policy_document" "analytical_dataset_write_s3" {
...
+      "kms:Encrypt",
+      "kms:Decrypt",
+      "kms:ReEncrypt*",
+      "kms:GenerateDataKey*",
+      "kms:DescribeKey",
+      "kms:CreateGrant"
     ]
 
     resources = [
-      "arn:aws:s3:::*",
```
[the remark this answers is not in the document]
Comment (UdayChokkam) [reply resources.tf]: Done

[the remark this answers is not in the document]
Comment (UdayChokkam) [reply resources.tf]: Merged all to one statement

File: resources.tf [hunk 9]
```diff
@@ -207,6 +327,100 @@ data "aws_iam_policy_document" "analytical_dataset_write_s3" {
...
+      "ssmmessages:OpenControlChannel",
+      "ssmmessages:OpenDataChannel"
+    ]
+    resources = ["*"]
+  }
+
+  statement {
+    effect = "Allow"
+    actions = [
+      "ec2messages:AcknowledgeMessage",
```
[the remark this answers is not in the document]
Comment (UdayChokkam) [reply resources.tf]: Done

File: resources.tf [hunk 10]
```diff
@@ -207,6 +327,100 @@ data "aws_iam_policy_document" "analytical_dataset_write_s3" {
...
+resource aws_iam_role_policy amazon_ec2_role_for_ssm {
+  role   = aws_iam_role.analytical_dataset_generator.name
+  policy = data.aws_iam_policy_document.amazon_ec2_role_for_ssm.json
+}
+
+data aws_iam_policy_document amazon_ec2_role_for_ssm {
+  statement {
+    effect = "Allow"
+    actions = [
+      "ssm:DescribeAssociation",
```
[the remark this answers is not in the document]
Comment (UdayChokkam) [reply resources.tf]: done

File: resources.tf [hunk 11]
```diff
@@ -247,6 +461,27 @@ resource "aws_security_group_rule" "analytical_dataset_generation_ingress" {
   security_group_id = aws_security_group.analytical_dataset_generation.id
 }
 
+resource "aws_security_group_rule" "egress_https_to_vpc_endpoints" {
```
[the remark this answers is not in the document]
Comment (UdayChokkam) [reply resources.tf:243]: This is required to communicate to Glue and other aws services without which spark, Hive wasn't connecting to Glue

File: resources.tf (hunk 1 above)
Comment (UdayChokkam) [reply resources.tf:254]: As this emr code exists in a different repo which is aws-analytical-dataset-generation than in aws-internal-compute repo where vpc code is . we will have to this way as an alternative

File: resources.tf (hunk 2 above)
Comment (UdayChokkam) [reply resources.tf:198]: done

File: emr.tf (hunk 3 above)
Comment (UdayChokkam) [reply emr.tf]: Changed as suggested

File: local.tf [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  dks_port                        = 8443
+  common_tags = {
+    Environment  = local.environment
+    Application  = local.emr_cluster_name
+    CreatedBy    = "terraform"
+    Owner        = "dataworks platform"
+    Persistence  = "Ignore"
+    AutoShutdown = "False"
+  }
+  internet_proxy = data.terraform_remote_state.internet_egress.outputs.internet_proxy
```
[the remark this answers is not in the document]
Comment (UdayChokkam) [reply local.tf]: done

File: emr.tf (hunk 5 above)
Comment (UdayChokkam) [reply emr.tf:59]: Line 50 is a configuration for various Hadoop services like spark , yarn , Hive , Hbase

File: emr-setup.sh (hunk 4 above)
Comment (UdayChokkam) [reply emr-setup.sh:3]: Yeah its safer to exit at first error for future when someone works on this script

Comment (ChrisScottThomas) [reply emr-setup.sh:3]: Sorry, I meant all the `set -*` you have.  -x will be especially verbose when running the script.  IMO, There's no need for this when not testing.

File: variables.tf (hunk 6 above)
Comment (UdayChokkam) [reply variables.tf]: changed

File: emr.tf (hunk 5 above)
Comment (ChrisScottThomas) [reply emr.tf:59]: I've not explained myself well enough.  Why can't we use a `steps.json`, like we have for `configuration.json`, instead if writing out the json inside a terraform file?

Comment (UdayChokkam) [reply emr.tf:59]: Looks like terraform doesn't have support for steps json unlike configurations_json

File: emr-setup.sh (hunk 4 above)
Comment (UdayChokkam) [reply emr-setup.sh:3]: removed

**Pull request merged**